Simplified/NYPLRemoteViewController.m

```diff
@@ -92,26 +100,90 @@ - (void)load
     return;
   }
 
-  NSTimeInterval timeoutInterval = 30.0;
-  NSTimeInterval activityLabelTimer = 10.0;
-
-  NSURLRequest *const request = [NSURLRequest requestWithURL:self.URL
-                                                 cachePolicy:NSURLRequestUseProtocolCachePolicy
-                                             timeoutInterval:timeoutInterval];
-
   // show "slow loading" label after `activityLabelTimer` seconds
   self.activityIndicatorLabel.hidden = YES;
   [NSTimer scheduledTimerWithTimeInterval: activityLabelTimer target: self
                                  selector: @selector(addActivityIndicatorLabel:) userInfo: nil repeats: NO];
 
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wdeprecated-declarations"
-  // TODO: SIMPLY-2589 Replace with NSURLSession
-  self.connection = [[NSURLConnection alloc] initWithRequest:request delegate:self];
-#pragma clang diagnostic pop
-  self.data = [NSMutableData data];
-  
-  [self.connection start];
+  self.dataTask = [NYPLNetworkExecutor.shared execute:request
+                           completion:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
+
+    dispatch_async(dispatch_get_main_queue(), ^{
+      [self.activityIndicatorView stopAnimating];
+      self.activityIndicatorLabel.hidden = YES;
+      NSURLSessionDataTask *dataTaskCopy = self.dataTask;
+      self.dataTask = nil;
+
+      if ([response.MIMEType isEqualToString:@"application/vnd.opds.authentication.v1.0+json"]) {
+        self.reloadView.hidden = false;
+        [NYPLAccountSignInViewController requestCredentialsUsingExistingBarcode:([NYPLUserAccount sharedAccount].barcode) completionHandler:^{
+          [self load];
+        }];
+        return;
+      }
+
+      if (error) {
+        self.reloadView.hidden = NO;
+        NSDictionary<NSString*, NSObject*> *metadata = @{
+          @"remoteVC.URL": self.URL ?: @"none",
+          @"connection.currentRequest": dataTaskCopy.currentRequest ?: @"none",
+          @"connection.originalRequest": dataTaskCopy.originalRequest ?: @"none",
+        };
+        [NYPLErrorLogger logError:error
+                          context:@"RemoteViewController"
+                          message:@"Server-side api call (likely related to Catalog loading) failed"
+                         metadata:metadata];
+
+        return;
+      }
+
+      NSHTTPURLResponse *httpResponse;
+      if ([response isKindOfClass: [NSHTTPURLResponse class]]) {
+        httpResponse = (NSHTTPURLResponse *) response;
+      }
+
+      if (httpResponse.statusCode != 200) {
+        [self handleErrorResponse:httpResponse withData:data];
+        return;
```

Comment:
not necessarily, previously there were 2 callbacks:
- (void)connection:(NSURLConnection *)connection didFailWithError:(NSError *)error
- (void)connectionDidFinishLoading:(__attribute__((unused)) NSURLConnection *)connection

And I don't think both of them were called in case of an error

Reply:
well, what I meant is that this `if statusCode != 200` was in the "success" case of the old code (connectionDidFinishLoading:) and there we were falling through the rest of the  code: https://github.com/NYPL-Simplified/Simplified-iOS/blob/develop/Simplified/NYPLRemoteViewController.m#L195

Reply:
ohhh, I see, you're obviously right, I changed it back